added: `NonLinMPC` and `MovingHorizonEstimator` integration with DI.jl

franckgaga opened the pull request:
**Nonlinear MPC and moving horzion estimator now use `DifferentiationInterface.jl`**

Both `NonLinMPC` and `MovingHorizonEstimator` constructor now accept a `gradient` and `jacobian` keyword arguments, receiving a `AbstractADType` to switch the backend of the objective gradient and constraint Jacobians, respectively. Note that the nonlinear inequality and equality constraints will both use the backend provided by the `jacobian` argument. The possibility to provide a hessian will be added in a future PR (the LBFGS approximation is used inside `Ipopt` for now, as it was the case before)

By default, denses `AutoForwardDiff()` is use everywhere, except for `NonLinMPC` with non-`SingleShooting` transcription method, in which this backend is used:

```julia
AutoSparse(
    AutoForwardDiff(); 
    sparsity_detector  = TracerSparsityDetector(), 
    coloring_algorithm = GreedyColoringAlgorithm()
)
```

For both objects, the differentiation make use of `DifferentiationInterface.Constant()` and `DifferentiationInterface.Cache()` functionalities. As a consequence, they does not rely on `PreallocationTools.DiffCache`s anymore. But, I still need to keep this dependency since `linearize!` and all `DiffSolver`s still rely on them for now.

**Documentation now use `DocumenterInterLink.jl`**

Some URL were very long in the docstrings. The are now way shorter. It also ease the maintenance for the documentation of external dependencies.

The doc preview is here: https://juliacontrol.github.io/ModelPredictiveControl.jl/previews/PR174

franckgaga committed:
added: `NonLinMPC` and `MovingHorizonEstimator` integration with DI.jl

franckgaga committed:
debug: typo in MHE

franckgaga committed:
starting support of sparse Jacobians

franckgaga committed:
WIP: using `Cache` instead of `DiffCache`

franckgaga committed:
added: compat on `DI.jl`

franckgaga committed:
Merge branch 'main' into diff_interface_final

franckgaga committed:
Merge branch 'debug_ms_terminal_2' into diff_interface_final

franckgaga committed:
Merge branch 'main' into diff_interface_final

franckgaga committed:
wip: support for sparse jacobians

franckgaga committed:
doc: long url separated from the text

franckgaga committed:
docs: making use of `DocumenterInterLink`

franckgaga committed:
doc: clean up in the inter-links

franckgaga committed:
added: use `DI.Cache` in `MovingHorizonEstimator`

franckgaga committed:
added: init diff. matrix as dense or sparse as required by backend

franckgaga committed:
debug: correct sparsity detection for `MovingHorizonEstimator`

franckgaga committed:
added: `mpc` and `estim` object as `DI.Constant`

franckgaga committed:
added: using `strict=Val(true)` un DI.jl preparation

franckgaga committed:
doc: replace urls with inter-links

franckgaga committed:
debug: `NonLinMPC` with `LinModel` and custom constraints now works

franckgaga committed:
changed: `update_prediction!` methods are now not nested
It is no longer necessary with `DI.Constant` context.

franckgaga committed:
test: remove useless test with new DI.jl interface

franckgaga commented:
Thanks for the info, right now I do:

```julia
mycontext = (
    Constant(a),
    Cache(b),
    Cache(c),
    Cache(d),
)
grad_prep = prepare_gardient!(myfunc, grad_backend, x, mycontext...; strict=Val(true))
```
and a similar splatting for the context in `gradient!`. Do you this splatting is problematic for the performance ? I did some quick benchamarks and I measured no impact at all.

gdalle commented:
I don't think it is problematic, but for some users it is a bit painful: when you have tens of caches you may want to structure the more neatly, so I give that option too.

franckgaga committed:
test: re-add the not so useless tests for coverage

franckgaga committed:
changed: remove unreachable branch in `MovingHorizonEstimator`

franckgaga committed:
bump

franckgaga committed:
doc: DI.jl mention in doc homepage

franckgaga committed:
doc: idem

franckgaga committed:
test: debug nmpc test

franckgaga committed:
test: new test NMPC and MHE tests with `FiniteDiff`

franckgaga committed:
added: store `AbstractADType` backends as struct parameters
This is "safer" like that. And it allows the user to verify what are the DI backends in a given MPC/MHE object. Also useful for the tests.

franckgaga committed:
test: new tests with `AutoFiniteDiff`
debug: temporarily fake a "filled" window for MHE gradient preparation (needed for some backend like FiniteDiff)

franckgaga committed:
debug: ambiguous methods in MHE

franckgaga commented:
@baggepinnen It's ready to be merged, if you have some comments. You can have a look a `NonLinMPC` and `MovingHorizonEstimator` documentation (add. details in their respective Extended Help)

franckgaga committed:
doc: MHE exetended help details on AD backends

gdalle commented:
> edit: I'm running the benchmarks of our papers in this version, and there is no significative difference on the performances, good news ❤️

If you want to play around some more, a parameter that might impact performance is the coloring algorithm used by SparseMatrixColorings, more specifically the [vertex ordering](https://gdalle.github.io/SparseMatrixColorings.jl/stable/api/#SparseMatrixColorings.AbstractOrder). Feel free to switch the default `NaturalOrder()` and see if it makes a difference with `SMC.ncolors(jacobian_prep)`. Roughly speaking, the sparse autodiff complexity scales with the number of colors $c$ (or with $\lceil c / 12 \rceil$ for ForwardDiff).

franckgaga commented:
@gdalle I just read that for `Constant()` context some backend requires `Number` or `AbstractArray`. 

Right now I pass a `mpc::NonLinMPC` object in the `Constant` arg. I understand that it won't work it some backend ? Do you know which one ?

gdalle commented:
I think it will only be symbolic backends. I should update the docstring. Not a hard limitation either, I just didn't get around to it.

gdalle commented:
> But it does not support Cache neither, and I use them extensively also (with normal Vectors).

Did you find a bug there? I think the docs might be out of date, `Cache`s with numbers or arrays should be supported by Symbolics too:

```julia
julia> using DifferentiationInterface

julia> import Symbolics

julia> f(x, c) = sum(copyto!(c, x))
f (generic function with 1 method)

julia> gradient(f, AutoSymbolics(), [1.0], Cache([0.0]))
1-element Vector{Int64}:
 1
```

gdalle commented:
> Okay, I just tested it and yes AutoSymbolics does not work with non-AbstractArray in Constant.

The reason why symbolic backends (and ReverseDiff) are more difficult for handling constants is that they generate an executable (or tape) for the function during preparation. If we close over the constants during preparation, they will be hardcoded into the executable. Therefore, we need to turn the constants into symbolic variables as well, so that the user is free to give different values at runtime.

I guess theoretically we could have a separation between "constant" `Constant`s (those which are the same at preparation and execution) and "variable" `Constant`s (those which differ between preparation and execution). But you can probably see how that would be confusing for the casual user, unless we come up with better names ^^

gdalle commented:
> > But it does not support Cache neither, and I use them extensively also (with normal Vectors).
> 
> Did you find a bug there? I think the docs might be out of date, `Cache`s with numbers or arrays should be supported by Symbolics too:
> 
> ```julia
> julia> using DifferentiationInterface
> 
> julia> import Symbolics
> 
> julia> f(x, c) = sum(copyto!(c, x))
> f (generic function with 1 method)
> 
> julia> gradient(f, AutoSymbolics(), [1.0], Cache([0.0]))
> 1-element Vector{Int64}:
>  1
> ```

Edit: I remember now, I marked `Cache`s as non-supported by Symbolics because tests fail for the Jacobian only:

https://github.com/JuliaDiff/DifferentiationInterface.jl/blob/cfab84dd775f6b7e9b39d0d0f84f0d260fc4fe09/DifferentiationInterface/test/Back/SymbolicBackends/symbolics.jl#L22-L27

franckgaga commented:
> > Okay, I just tested it and yes AutoSymbolics does not work with non-AbstractArray in Constant.
> 
> The reason why symbolic backends (and ReverseDiff) are more difficult for handling constants is that they generate an executable (or tape) for the function during preparation. If we close over the constants during preparation, they will be hardcoded into the executable. Therefore, we need to turn the constants into symbolic variables as well, so that the user is free to give different values at runtime.
> 
> I guess theoretically we could have a separation between "constant" `Constant`s (those which are the same at preparation and execution) and "variable" `Constant`s (those which differ between preparation and execution). But you can probably see how that would be confusing for the casual user, unless we come up with better names ^^

Yeah, this "nomenclature issue" crossed my mind during the implementation. And I literally use both "defintion" in the package right now: `Constant(mpc)` is literally a constant that won't change at runtime.

In the `linearize` implementation of #179, the `Constant`s are vectors that are not mutated during the differentiation but their value is different at each `jacobian!` call.

Worst idea ever: a new context called `ConstantConstant()` :laughing:

franckgaga commented:
Okay, I'm a bit ambivalent on if should keep the `Constant(mpc)` or not. 

My feeling right now: since it does not need any modification at runtime (same object at each `jacobian!` call), it's a "constant" `Constant` and I can use the closure trick. Theoretically, if `AutoSymbolics` improve in the future, `Cache` with normal vectors will be well supported, and it will work well for `gradient`, `jacobian` etc.

Conclusion: I will remove `Constant(mpc)` (and same thing for MHE).

gdalle commented:
> My feeling right now: since it does not need any modification at runtime (same object at each jacobian! call), it's a "constant" Constant and I can use the closure trick.

I tend to agree with that. Essentially, if your constant never changes, it's part of your function, and then why should DI give a shit ^^

> Theoretically, if AutoSymbolics improve in the future, Cache with normal vectors will be well supported, and it will work well for gradient, jacobian etc.

Looking at it again, I think it is a bug in Symbolics. Will try to dig up an MWE

franckgaga committed:
changed: avoid non-`AbstractArray`s in `DI.Constant()`
By using a closure on `mpc::NonLinMPC` and `estim::MovingHorizonEstimator`

franckgaga commented:
@baggepinnen I will merge the PR, you can comment here if you spot anything not clear or sus

baggepinnen commented:
Did you end up getting things to work with Symbolics? If so, it would be nice to see if you get any benefits from the new common-subexpression elimination (CSE) optimization that was recently enabled. See https://github.com/JuliaDiff/DifferentiationInterface.jl/issues/758 for a discussion, I got 13x speedup on a sparse hessian example for a cartpole system. My guess is that this optimization will give the most benefit for mechanical systems with non-trivial mass matrix, e.g., 2DOF and higher

gdalle commented:
Unfortunately not, but here's an MWE if you have a few cycles to dig into why Symbolics fails. I think it is due to Symbolics because `jacobian` is the only operator that fails with `DI.Cache`, all the others work as intended:

```julia
using DifferentiationInterface
using Symbolics: Symbolics

function f!(y, x)
    y[1:length(x)] .= sin.(x)
    y[(length(x) + 1):(2length(x))] .= cos.(x)
    return nothing
end

function f!_cache(y, x, c)
    f!(c, x)
    copyto!(y, c)
    return nothing
end
```

```julia
julia> x = [1.0, 2.0];

julia> y, c = zeros(4), zeros(4);

julia> jacobian(f!, y, AutoSymbolics(), x)  # correct
4×2 Matrix{Float64}:
  0.540302   0.0
  0.0       -0.416147
 -0.841471   0.0
  0.0       -0.909297

julia> jacobian(f!_cache, y, AutoSymbolics(), x, Cache(c))  # incorrect
4×2 Matrix{Float64}:
  0.0   0.0
  0.0   0.0
 -0.0   0.0
  0.0  -0.0

gdalle commented:
Here's a pure Symbolics MWE: https://github.com/JuliaSymbolics/Symbolics.jl/issues/1509

baggepinnen commented:
The issue is that `context_vars` is overwritten when computing `f!(y_var, x_var, context_vars...)` so that 
```julia
context_vars =
Vector{Symbolics.Num}[[sin(x₁), sin(x₂), cos(x₁), cos(x₂)]]
```
when the funciton is later built, it looks like the output is a function of those `sin(x₁), sin(x₂), cos(x₁), cos(x₂)` inputs, but when cache is passed as an array of zeros you get zeros out. 

I don't think the symbolics backend will benefit from a cache, it will trace through the function and build up an expression that will be used to symbolically compute the jacobian. this tracing does not care about any intermediate allocations while building the expression, since those will not be represented in the expression once it has been built

baggepinnen commented:
Another way of thinking about it is that Symbolics cannot represent _computation_, it can only represent _expressions_. The final expression for the output is the same no matter whether the cache is used as an intermediate or not.

baggepinnen commented:
If you call `f_cache` once before such that `c` contains the correct intermediate _values_, you do get the correct jacobian out
```julia
julia> f!_cache(y, x, c)

julia> jacobian(f!_cache, y, AutoSymbolics(), x, Cache(c))
4×2 Matrix{Float64}:
  0.540302   0.0
  0.0       -0.416147
 -0.841471   0.0
  0.0       -0.909297
```
but there is no benefit in using such a cumbersome approach, the built jacobian funciton should be allocation free even if no cache was used during the symbolic tracing

gdalle commented:
So I guess the solution would be to build the function based solely on `x` and any constant contexts, but not `cache` contexts?

baggepinnen commented:
> So I guess the solution would be to build the function based solely on x and any constant contexts, but not cache contexts?

Yes, I think this should be fine. You might need to wrap the built function such that it takes the correct number of arguments if that's expected downstream. Alternatively, you could recreate `context_vars` such that they are once again given by `cache_1, cache_2...` and then symbolics will not find them in the output and you'd get the function computing everything in the output from `x_vars` instead of from `cache_vars`

gdalle commented:
Fix incoming in https://github.com/JuliaDiff/DifferentiationInterface.jl/pull/760 but there is still something I don't understand.
Let's say I run a function `f!(y, x, c)` and at the end, `c = sin(x[1])` and `y = cos(c[1]) = cos(sin(x[1]))`. How does Symbolics decide that it will be _defining_ `y` based on `c` instead of `x`?

baggepinnen commented:
I am not sure, but my _guess_ is that there is a tree search on the output expression taking place, and the first subexpression that matches any of the input arguments will be used, and when exploring the tree, `sin(x[1])` appears closer to the root than `x[1]`

franckgaga commented:
> Fix incoming in [JuliaDiff/DifferentiationInterface.jl#760](https://github.com/JuliaDiff/DifferentiationInterface.jl/pull/760) but there is still something I don't understand. Let's say I run a function `f!(y, x, c)` and at the end, `c = sin(x[1])` and `y = cos(c[1]) = cos(sin(x[1]))`. How does Symbolics decide that it will be _defining_ `y` based on `c` instead of `x`?

Just tested `AutoSymbolics()` and it works perfectly now, good job @gdalle! :partying_face: :partying_face: :partying_face: 

I will do some benchmarks to see if it's faster than `AutoForwardDiff()` in the sparse case.

gdalle commented:
I'm rather curious!

franckgaga commented:
The benchmarks on the inverted pendulum with `MultipleShooting` and `Ipopt` are:
```julia
# dense AutoForwardDiff and MultipleShooting:
btime_NMPC_track_solver_IP = median(bm) = TrialEstimate(1.253 s)

# dense AutoSymbolics and MultipleShooting:
btime_NMPC_track_solver_IP = median(bm) = TrialEstimate(1.077 s)

# sparse AutoForwardDiff and MultipleShooting:
btime_NMPC_track_solver_IP = median(bm) = TrialEstimate(1.269 s)
```
So even on this really small and simple ODE with a trivial mass matrix (but unstable) there is a gain with `AutoSymbolics`.

Now, I'm not able to run it with sparse `AutoSymbolics`:
```julia
backend = AutoSparse(
    AutoSymbolics(); 
    sparsity_detector  = TracerSparsityDetector(), 
    coloring_algorithm = GreedyColoringAlgorithm()
)
```
the error is:
```julia
ERROR: MethodError: no method matching sparsity_pattern(::DifferentiationInterfaceSymbolicsExt.SymbolicsTwoArgJacobianPrep{…})
The function `sparsity_pattern` exists, but no method is defined for this combination of argument types.

Closest candidates are:
  sparsity_pattern(::DifferentiationInterfaceSparseMatrixColoringsExt.SparseJacobianPrep)
   @ DifferentiationInterfaceSparseMatrixColoringsExt ~/.julia/packages/DifferentiationInterface/Av6Jb/ext/DifferentiationInterfaceSparseMatrixColoringsExt/DifferentiationInterfaceSparseMatrixColoringsExt.jl:20
  sparsity_pattern(::DifferentiationInterfaceSparseMatrixColoringsExt.SparseHessianPrep)
   @ DifferentiationInterfaceSparseMatrixColoringsExt ~/.julia/packages/DifferentiationInterface/Av6Jb/ext/DifferentiationInterfaceSparseMatrixColoringsExt/hessian.jl:21
  sparsity_pattern(::AbstractColoringResult)
   @ SparseMatrixColorings ~/.julia/packages/SparseMatrixColorings/BED6Y/src/result.jl:127

Stacktrace:
 [1] init_diffmat(T::Type, backend::AutoSparse{…}, prep::DifferentiationInterfaceSymbolicsExt.SymbolicsTwoArgJacobianPrep{…}, ::Int64, ::Int64)
   @ ModelPredictiveControl ~/.julia/dev/ModelPredictiveControl/src/general.jl:61
 [2] get_optim_functions(mpc::NonLinMPC{…}, ::Model)
   @ ModelPredictiveControl ~/.julia/dev/ModelPredictiveControl/src/controller/nonlinmpc.jl:659
 [3] init_optimization!(mpc::NonLinMPC{…}, model::NonLinModel{…}, optim::Model)
   @ ModelPredictiveControl ~/.julia/dev/ModelPredictiveControl/src/controller/nonlinmpc.jl:543
 [4] (NonLinMPC{…})(estim::UnscentedKalmanFilter{…}, Hp::Int64, Hc::Int64, M_Hp::Matrix{…}, N_Hc::Matrix{…}, L_Hp::Matrix{…}, Cwt::Float64, Ewt::Float64, JE::ModelPredictiveControl.var"#159#163", gc!::ModelPredictiveControl.var"#160#164", nc::Int64, p::Vector{…}, transcription::MultipleShooting, optim::Model, gradient::AutoForwardDiff{…}, jacobian::AutoSparse{…})
   @ ModelPredictiveControl ~/.julia/dev/ModelPredictiveControl/src/controller/nonlinmpc.jl:126
 [5] NonLinMPC(estim::UnscentedKalmanFilter{…}; Hp::Int64, Hc::Int64, Mwt::Vector{…}, Nwt::Vector{…}, Lwt::Vector{…}, M_Hp::Matrix{…}, N_Hc::Matrix{…}, L_Hp::Matrix{…}, Cwt::Float64, Ewt::Float64, JE::ModelPredictiveControl.var"#159#163", gc!::ModelPredictiveControl.var"#160#164", gc::ModelPredictiveControl.var"#160#164", nc::Int64, p::Vector{…}, transcription::MultipleShooting, optim::Model, gradient::AutoForwardDiff{…}, jacobian::AutoSparse{…})
   @ ModelPredictiveControl ~/.julia/dev/ModelPredictiveControl/src/controller/nonlinmpc.jl:393
 [6] top-level scope
   @ ~/Dropbox/Programmation/Julia/TestMPC/src/nonlin_mpc_article.jl:52
Some type information was truncated. Use `show(err)` to see complete types.
```
Is it expected ?

gdalle commented:
No this is not expected but I'm not sure I can solve it either. Essentially, I overload analysis functions like `SMC.sparsity_pattern` or `SMC.ncolors` on generic sparse preparation result types. The thing is, Symbolics doesn't take the usual sparse AD pipeline, so there won't be a coloring to anlayze in the first place (and the sparsity pattern isn't computed explicitly either). If you want the sparsity pattern, you'll have to get it separately with `ADTypes.jacobian_sparsity` and `Symbolics.SymbolicsSparsityDetector()`.

gdalle commented:
Can you open an issue in the DI repo for me to make the error more informative?

franckgaga commented:
see https://github.com/JuliaDiff/DifferentiationInterface.jl/issues/763

gdalle commented:
Could you benchmark against sparse FastDifferentiation too?

franckgaga commented:
> The benchmarks on the inverted pendulum with `MultipleShooting` and `Ipopt` are:
> 
> ```julia
> # dense AutoForwardDiff and MultipleShooting:
> btime_NMPC_track_solver_IP = median(bm) = TrialEstimate(1.253 s)
> 
> # dense AutoSymbolics and MultipleShooting:
> btime_NMPC_track_solver_IP = median(bm) = TrialEstimate(1.077 s)
> 
> # sparse AutoForwardDiff and MultipleShooting:
> btime_NMPC_track_solver_IP = median(bm) = TrialEstimate(1.269 s)
> ```

So with the new bugfix related sparse symbolic backends, I get:

```julia
# sparse AutoSymbolics and MultipleShooting:
btime_NMPC_track_solver_IP = median(bm) = TrialEstimate(1.045 s)
```
Thus not much difference with dense computations on this problem. Note that the bottleneck may be Ipopt here. If that's true even faster differentiation algorithms would not change much the result. But I would need to do some profiling to verify this.

For `FastDifferentiation`, both dense and sparse backend does not work. The error on the dense case is:
```julia
ERROR: AssertionError: Should only be one path from root 5 to variable 1. Instead have 2 children from node 122 on the path
Stacktrace:
  [1] follow_path(a::FastDifferentiation.DerivativeGraph{Int64}, root_index::Int64, var_index::Int64)
    @ FastDifferentiation ~/.julia/packages/FastDifferentiation/kJ1qL/src/Factoring.jl:525
  [2] evaluate_path
    @ ~/.julia/packages/FastDifferentiation/kJ1qL/src/Factoring.jl:555 [inlined]
  [3] _symbolic_jacobian!(graph::FastDifferentiation.DerivativeGraph{…}, partial_variables::Vector{…})
    @ FastDifferentiation ~/.julia/packages/FastDifferentiation/kJ1qL/src/Jacobian.jl:47
  [4] _symbolic_jacobian
    @ ~/.julia/packages/FastDifferentiation/kJ1qL/src/Jacobian.jl:61 [inlined]
  [5] jacobian
    @ ~/.julia/packages/FastDifferentiation/kJ1qL/src/Jacobian.jl:90 [inlined]
  [6] prepare_jacobian_nokwarg(::Val{…}, ::Function, ::Vector{…}, ::AutoFastDifferentiation, ::Vector{…}, ::Cache{…}, ::Cache{…}, ::Cache{…}, ::Cache{…}, ::Cache{…}, ::Cache{…}, ::Cache{…}, ::Cache{…}, ::Cache{…}, ::Cache{…}, ::Cache{…})
    @ DifferentiationInterfaceFastDifferentiationExt ~/.julia/packages/DifferentiationInterface/7eD1K/ext/DifferentiationInterfaceFastDifferentiationExt/twoarg.jl:320
  [7] #prepare_jacobian#48
    @ ~/.julia/packages/DifferentiationInterface/7eD1K/src/first_order/jacobian.jl:23 [inlined]
  [8] prepare_jacobian
    @ ~/.julia/packages/DifferentiationInterface/7eD1K/src/first_order/jacobian.jl:15 [inlined]
  [9] get_optim_functions(mpc::NonLinMPC{…}, ::JuMP.Model)
    @ ModelPredictiveControl ~/.julia/dev/ModelPredictiveControl/src/controller/nonlinmpc.jl:706
 [10] init_optimization!(mpc::NonLinMPC{…}, model::NonLinModel{…}, optim::JuMP.Model)
    @ ModelPredictiveControl ~/.julia/dev/ModelPredictiveControl/src/controller/nonlinmpc.jl:543
 [11] (NonLinMPC{…})(estim::UnscentedKalmanFilter{…}, Hp::Int64, Hc::Int64, M_Hp::Matrix{…}, N_Hc::Matrix{…}, L_Hp::Matrix{…}, Cwt::Float64, Ewt::Float64, JE::ModelPredictiveControl.var"#159#163", gc!::ModelPredictiveControl.var"#160#164", nc::Int64, p::Vector{…}, transcription::MultipleShooting, optim::JuMP.Model, gradient::AutoForwardDiff{…}, jacobian::AutoFastDifferentiation)
    @ ModelPredictiveControl ~/.julia/dev/ModelPredictiveControl/src/controller/nonlinmpc.jl:126
 [12] NonLinMPC(estim::UnscentedKalmanFilter{…}; Hp::Int64, Hc::Int64, Mwt::Vector{…}, Nwt::Vector{…}, Lwt::Vector{…}, M_Hp::Matrix{…}, N_Hc::Matrix{…}, L_Hp::Matrix{…}, Cwt::Float64, Ewt::Float64, JE::ModelPredictiveControl.var"#159#163", gc!::ModelPredictiveControl.var"#160#164", gc::ModelPredictiveControl.var"#160#164", nc::Int64, p::Vector{…}, transcription::MultipleShooting, optim::JuMP.Model, gradient::AutoForwardDiff{…}, jacobian::AutoFastDifferentiation)
    @ ModelPredictiveControl ~/.julia/dev/ModelPredictiveControl/src/controller/nonlinmpc.jl:393
 [13] top-level scope
    @ ~/Dropbox/Programmation/Julia/TestMPC/src/nonlin_mpc_article.jl:52
Some type information was truncated. Use `show(err)` to see complete types.
```
The error is different on the sparse case. I can post it also here if you are interested @gdalle, but I think it's not relevant for now since it first needs to work in the dense case before it starts working in the sparse case, logically.

gdalle commented:
Could you post a full MWE in a DI issue? I think this is related to cache handling in FastDifferentiation, probably a similar error as the one you diagnosed earlier with Symbolics

franckgaga commented:
I will try to create a MWE.

franckgaga commented:
I don't think it's related to cache handling since this MWE with cache work as expected:
```julia
using DifferentiationInterface
import FastDifferentiation
x = float.(1:8)
y = Vector{eltype(x)}(undef, length(x)-1)
c = Cache(y)
function f!(y, x, c)
    for i in eachindex(y)
        c[i] = (x[i+1]^2-x[i]^2)
    end
    for i in eachindex(y)
        j = length(x) - i
        c[i] += (x[j]^2-x[j+1]^2)
    end
    y .= c
    return y
end
backend = AutoFastDifferentiation()
prep = prepare_jacobian(f!, y, backend, x, c; strict=Val(true))
J = jacobian(f!, y, prep, backend, x, c)
```
giving:
```julia
7×8 Matrix{Float64}:
 -2.0   4.0   0.0   0.0    0.0    0.0   14.0  -16.0
  0.0  -4.0   6.0   0.0    0.0   12.0  -14.0    0.0
  0.0   0.0  -6.0   8.0   10.0  -12.0    0.0    0.0
  0.0   0.0   0.0   0.0    0.0    0.0    0.0    0.0
  0.0   0.0   6.0  -8.0  -10.0   12.0    0.0    0.0
  0.0   4.0  -6.0   0.0    0.0  -12.0   14.0    0.0
  2.0  -4.0   0.0   0.0    0.0    0.0  -14.0   16.0
```
I think we are hitting some language constructs not supported by `FastDifferentiation`. I have some `for` loops and ìf` conditionals in the nonlinear objective and nonlinear constraint functions. I can still raise an issue on DI.jl with a not-so-minimal reproductive exemple that uses MPC.jl, if you are interested.
